// @sap/eslint-plugin-cds/lib/rules/java/cql-class-targets.js
'use strict'

/**
 * In Select.from(x), x should be a class literal.
 */

const { RULE_CATEGORIES } = require('../../constants')

module.exports = {
  meta: {
    type: 'problem',
    docs: {
      recommended: true,
      category: RULE_CATEGORIES.java,
      description: 'java.'
    },
    fixable: 'code',
    schema: [],
    messages: {
      selectOnNonClass: `'{{target}}' is not a class. Prefer to pass class literals to Select.from!`,
    },
    hasSuggestions: true
  },
  create: context => {
    const knownClasses = new Set()

    /** @param {Node} node */
    // const refersToClass = node =>
    //   node.type === 'class_literal'
    //   // or part of explicit named imports
    //   || knownClasses.has(node.text)
    //   // or first letter of last part of text is uppercase: a.b.C -> true
    //   // https://unicode.org/reports/tr18/#General_Category_Property
    //   || /^\p{Lu}/u.test(node.text.split('.').at(-1))

    const isString = node =>
      node?.type === 'string_literal'

    return {
      'import_declaration > scoped_identifier': function(node) {
        if (!node) return
        if (node.nameNode) {
          knownClasses.add(node.nameNode.text)
        }
      },

      "method_invocation > identifier[text='from']": function(node) {
        if (!node?.parent?.argumentsNode) return
        const args = node.parent.argumentsNode.children
          .filter(c => c.constructor.name !== 'SyntaxNode')  // filter out commas, parens, etc.
          ?? []
        // we actually want to detect anything that is NOT a class literal,
        // but there are some edge cases we need to care for in the future,
        // such as parameters passing in a class:
        // void myMethod(Class<MyClass> cls) { Select.from(cls); }
        // for the time being, we will just warn on string literals
        //if (!refersToClass(args[0])) {
        if (isString(args[0])) {
          context.report({
            node: args[0],
            messageId: 'selectOnNonClass',
            data: { target: args[0].text }
          })
        }
      },
    }
  }
}
