// @sap/eslint-plugin-cds/lib/languages/java/java-language.js
'use strict'

const { TextEncoder } = require('node:util')
const TreeSitterParser = require('tree-sitter')
const Java = require('tree-sitter-java')
const { JavaSourceCode } = require('./java-source-code')

/**
 * @typedef {object} JavaLanguageOptions
 * @typedef {{languageOptions: JavaLanguageOptions}} JavaLanguageContext
 */

const jp = new TreeSitterParser()
jp.setLanguage(Java)

/**
 * @param {import('@eslint/core').File} file
 * @param {JavaLanguageContext} context
 */
function parseJavaCode (file /*, context*/) {
  const code = file.body
  try {
    // https://stackoverflow.com/a/79771252
    const codeByteSize = new TextEncoder().encode(code).length
    const treeSitterAst = jp.parse(code, undefined, { bufferSize: codeByteSize + 32 })  // adding some extra buffer space
    if (!treeSitterAst) return { ast: null, ok: false, errors: [] }
    if (!treeSitterAst.rootNode) return { ast: treeSitterAst, ok: false, errors: [] }
    return { ast: treeSitterAst, ok: true, errors: [] }
  } catch (e) {
    // most likely a parsing error. This could mean we have to update the parser for newer Java versions
    return { ast: null, ok: false, errors: [e] }
  }
}

/**
 * @param {import('@eslint/core').File} file
 * @param {ReturnType<typeof parseJavaCode>} parseResult
 * @param {JavaLanguageContext} context
 */
function createJavaSourceCode(file, parseResult /*, context*/) {
  if (!parseResult.ok || !parseResult.ast) {
    throw new Error('createJavaSourceCode: parseResult is not ok or has no AST', {
      ok: parseResult.ok,
      hasAst: !!parseResult.ast
    })
  }
  return new JavaSourceCode({ text: file.body, ast: parseResult.ast })
}

/**
 * @param {JavaLanguageOptions} languageOptions
 */
function validateJavaLanguageOptions(/*languageOptions*/) {
  // No specific options to validate for Java currently
}

/**
 * @link https://eslint.org/docs/latest/extend/languages#the-language-object
 */
const javaLanguage = /** @type {const}*/({
  fileType: 'text',
  lineStart: 0,
  columnStart: 0,
  nodeTypeKey: 'type',
  validateLanguageOptions: validateJavaLanguageOptions,
  parse: parseJavaCode,
  createSourceCode: createJavaSourceCode
})

module.exports = {
  javaLanguage,
}
