// @sap/eslint-plugin-cds/lib/languages/java/java-source-code.js
'use strict'

// not 100% accurate, as it matches _any_ comment containing 'eslint-disable' etc.
// but in Java sources, users will hopefully only mention eslint-* when they
// actually mean to configure ESLint.
const INLINE_CONFIG = /eslint(?:-enable|-disable(?:(?:-next)?-line)?)?(?:\s|$)/u

/**
 * @link see https://github.com/eslint/json/blob/main/src/languages/json-source-code.js
 */
const {
  VisitNodeStep,
  TextSourceCodeBase,
  Directive
} = require('@eslint/plugin-kit')

/** @typedef {'disable' | 'enable' | 'disable-next-line' | 'disable-line'} DirectiveType */


/**
 * @param {import('tree-sitter').SyntaxNode} comment
 * @returns {{
 *  label: 'eslint-disable' | 'eslint-enable' | 'eslint-disable-next-line' | 'eslint-disable-line',
 *  value: string,
 *  justification: string | undefined
 * }}
 */
function parseDirective(comment) {
  // label, value, justification
  //      * but before the "--" that indicates the justification.
  const [, label, value, justification] = comment.text.match(/^\W*(eslint[\w-]*) (\S+)(?: -- (.+))?/)
  return { label, value, justification}
}

class JavaTraversalStep extends VisitNodeStep {
  /** @param {{ target: unknown, phase: 1|2, args: Array<any> }} arg0 */
  constructor({ target, phase, args }) {
    super({ target, phase, args })
  }
}

/**
 * @param {import('tree-sitter').SyntaxNode} node
 * @param {(node: import('tree-sitter').SyntaxNode) => void} enterAction
 * @param {undefined | (node: import('tree-sitter').SyntaxNode) => void} exitAction
 */
function traverse(node, enterAction, exitAction) {
  enterAction?.(node)
  //steps.push(new JavaTraversalStep({ target: node, phase: 1, args: [] }))
  for (const child of node.children) {
    traverse(child, enterAction, exitAction)
  }
  exitAction?.(node)
}


/** @param {import('tree-sitter').SyntaxNode} node */
const getAncestors = node => !node
  ? []
  : [...getAncestors(node.parent), node]

/**
 * @link https://eslint.org/docs/latest/extend/languages#the-sourcecode-object
 */
class JavaSourceCode extends TextSourceCodeBase {
  /** @type {undefined | import('tree-sitter').SyntaxNode[]} */
  #comments
  /** @type {undefined | import('tree-sitter').SyntaxNode[]} */
  #inlineConfigComments

  get comments() {
    if (!this.#comments) {
      this.#comments = []
      traverse(this.ast.rootNode, node => {
        if (['line_comment', 'block_comment'].includes(node.type)) {
          this.#comments.push(node)
        }
      })
    }
    return this.#comments
  }

  get inlineConfigNodes() {
    if (!this.#inlineConfigComments) {
      this.#inlineConfigComments = this
        .comments
        .filter(comment =>
          INLINE_CONFIG.test(comment.text),
        ) ?? []
    }

    return this.#inlineConfigComments
  }

  get lines () { return this.text.split(/\r\n|\r|\n/)  }

  /** @param {import('tree-sitter').SyntaxNode} node */
  getLoc(node) {
    return {
      start: { line: node.startPosition.row, column: node.startPosition.column },
      end: { line: node.endPosition.row, column: node.endPosition.column }
    }
  }

  /** @param {import('tree-sitter').SyntaxNode} node */
  getRange(node) {
    return [node.startIndex, node.endIndex]
  }

  /** @param {import('tree-sitter').SyntaxNode} node */
  getParent(node) {
    return node.parent ?? undefined  // .parent is null if root, we need undefined
  }

  /** @param {import('tree-sitter').SyntaxNode} node */
  getAncestors(node) {
    // ESLint wants ancestors excluding the node itself
    return getAncestors(node.parent)
  }

  /** @param {import('tree-sitter').SyntaxNode} node */
  getText(node) {
    return node.text
  }

  getDisableDirectives() {
    /** @type {FileProblem[]} */
    const problems = []
    /** @type {Directive[]} */
    const directives = []

    for (const comment of this.inlineConfigNodes) {
      const { label, value, justification } = parseDirective(comment)

      // `eslint-disable-line` directives are not allowed to span multiple lines as it would be confusing to which lines they apply
      if (label === 'eslint-disable-line' && comment.loc.start.line !== comment.loc.end.line) {
        problems.push({
          ruleId: null,
          message: `${label} comment should not span multiple lines.`,
          loc: comment.loc,
        })
      } else if (['eslint-disable', 'eslint-enable', 'eslint-disable-next-line', 'eslint-disable-line'].includes(label)) {
        directives.push(
          new Directive({
            type: /** @type {DirectiveType} */ (label.slice('eslint-'.length)),
            node: comment,
            value,
            justification,
          }),
        )
      }
    }
    return { problems, directives }
  }

  traverse() {
    const steps = []
    if (!this.ast || !this.ast.rootNode) {
      return steps
    }
    traverse(this.ast.rootNode, node => {
      steps.push(new JavaTraversalStep({ target: node, phase: 1 }))
    }, node => {
      steps.push(new JavaTraversalStep({ target: node, phase: 2 }))
    })
    return steps
  }

  /** @param {{text: string, ast: import('tree-sitter').Tree}} param0 */
  constructor({text, ast}) {
    super({text, ast})
    this.text = text
    this.ast = ast
  }
}

module.exports = {
  JavaSourceCode
}
