// @sap/eslint-plugin-cds/lib/rules/assoc2many-ambiguous-key.js
'use strict'

const cds = require('@sap/cds')

/** @type {import('../types').Rule} */

const isAssociationOrComposition = element => ['cds.Association', 'cds.Composition'].includes(element?.type)
const isEntity = element => element?.kind === 'entity'
const isStarCardinality = element => element?.cardinality?.max === '*'
/**
 * Using infix filters, one-to-many associations can be reduced to one-to-one.
 * @example
 * ```cds
 * SELECT name, books[1: favorite=true].title from Authors
 * ```
 */
const isScalarInfixFilter = element => element?.ref?.[0]?.cardinality?.max === 1

function resolveSourceAlias (csn, definition) {
  const from = definition.query.SELECT.from
  let sourceEntity
  const sourceAlias = []
  if (from?.ref) {
    // From
    sourceEntity = csn.definitions[from.ref.join('_')]
    sourceAlias.push({
      from: sourceEntity.name,
      as: from.as ?? from.ref.at(-1).split('.').pop()
    })
  } else if (from?.args?.[0].ref) {
    // Join
    sourceEntity = csn.definitions[from.args[0].ref.join('_')]
    for (const arg of from.args) {
      sourceAlias.push({
        from: arg.ref.join('_'),
        as: arg.as ?? arg.ref.at(-1).split('.').pop()
      })
    }
  }
  return { sourceEntity, sourceAlias }
}

module.exports = {
  meta: {
    schema: [{/* to avoid deprecation warning for ESLint 9 */}],
    docs: {
      description: 'Ambiguous key with a `TO MANY` relationship since entries could appear multiple times with the same key.',
      category: 'Model Validation',
      recommended: true,
      url: 'https://cap.cloud.sap/docs/tools/cds-lint/rules/assoc2many-ambiguous-key',
    },
    messages: {
      ambiguous: `Ambiguous key in '{{name}}'. Element '{{column-name}}' leads to multiple entries so that key '{{key-name}}' is not unique.`,
    },
    type: 'problem',
    model: 'inferred'
  },
  create (context) {
    return checkAssocs

    function checkAssocs () {
      let csnOdata
      const m = context.getModel()
      if (m?.definitions) {
        try {
          csnOdata = cds.compile.for.odata(m)
          const csnOdataLinked = cds.linked(csnOdata)
          associationCardinalityFlaw(csnOdataLinked, context)
        } catch {
          // FIXME: there are currently too many issues with this rule, e.g.
          //        assumptions that properties exist, etc.
        }
      }
    }
  }
}

/**
 * @param {import('../types').CSN} csn
 * @param {import('../types').CDSRuleContext} context
 */
function associationCardinalityFlaw (csn, context) {
  for (const [name, definition] of Object.entries(csn.definitions)) {
    if (!name.startsWith('localized.')) {
      processEntity({ definition, csn, context})
    }
  }
}

/**
 * @param {import('../types').CSN} csn
 */
function processEntity ({ definition, csn, context}) {
  if (!isEntity(definition) || !definition.query?.SELECT?.columns) return

  const { sourceEntity, sourceAlias } = resolveSourceAlias(csn, definition)
  if (!sourceEntity) return

  for (const column of definition.query.SELECT.columns) {
    processElement({ csn, definition, column, sourceEntity, sourceAlias, context })
  }
}

/**
 * @param {object} p
 * @param {import('../types').CSN} p.csn
 * @param {object} p.column
 * @param {object} p.definition
 * @param {object} p.sourceEntity
 * @param {string} p.sourceAlias
 * @param {import('../types').CDSRuleContext} p.context
 */
function processElement ({csn, definition, column, sourceEntity, sourceAlias, context}) {
  if (!(column.ref && column.ref.length > 1)) return

  let refPlainElement = false
  let refCardinalityMult = false
  let refEntity = sourceEntity
  const refAlias = sourceAlias
  for (const ref of column.ref.map(ref => ref.id ?? ref)) {
    // Alias
    const matchAlias = refAlias.find(alias => alias.as === ref)
    let refElement
    if (matchAlias) {
      refEntity = csn.definitions[matchAlias.from]
    } else {
      refElement = refEntity.elements[ref]
          // Mixin
          ?? definition.elements[ref]
          ?? definition.query.SELECT.mixin[ref]
          ?? definition.query.SELECT.mixin[column.ref[0]]?._target?.elements[ref]

      if (isAssociationOrComposition(refElement)) {
        refCardinalityMult = isStarCardinality(refElement)
      } else {
        refPlainElement = true
      }
    }

    if (isAssociationOrComposition(refElement)) {
      refEntity = csn.definitions[refElement.target]
    }
  }

  const keyNames = Object.keys(definition.keys ?? {})
  const [firstkey] = keyNames
  if (keyNames.length === 1 && firstkey === 'ID' && refCardinalityMult && refPlainElement && !isScalarInfixFilter(column)) {
    const key = definition.keys[firstkey]
    context.report({
      messageId: 'ambiguous',
      data: {
        name: definition.name,
        'column-name': column.as ?? column.name,
        'key-name': firstkey
      },
      loc: context.getLocation(firstkey, key, csn),
      file: key.$location.file
    })
  }
}
